subprojects/workers/src/main/java/org/gradle/workers/internal/DefaultWorkerExecutor.java

```diff
@@ -72,36 +83,67 @@ public DefaultWorkerExecutor(WorkerFactory daemonWorkerFactory, WorkerFactory is
         this.workerDirectoryProvider = workerDirectoryProvider;
         this.classLoaderStructureProvider = classLoaderStructureProvider;
         this.actionExecutionSpecFactory = actionExecutionSpecFactory;
+        this.instantiator = instantiator;
     }
 
     @Override
     public void submit(Class<? extends Runnable> actionClass, Action<? super WorkerConfiguration> configAction) {
         WorkerConfiguration configuration = new DefaultWorkerConfiguration(forkOptionsFactory);
-        File workingDirectory = workerDirectoryProvider.getWorkingDirectory();
-        configuration.getForkOptions().setWorkingDir(workingDirectory);
         configAction.execute(configuration);
-        String description = configuration.getDisplayName() != null ? configuration.getDisplayName() : actionClass.getName();
 
-        if (!workingDirectory.equals(configuration.getForkOptions().getWorkingDir())) {
+        Action<WorkerSpec<AdapterWorkerParameters>> action = new Action<WorkerSpec<AdapterWorkerParameters>>() {
+            @Override
+            public void execute(WorkerSpec<AdapterWorkerParameters> adapterWorkerParametersWorkerSpec) {
+                AdapterWorkerParameters parameters = adapterWorkerParametersWorkerSpec.getParameters();
+                parameters.setImplementationClassName(actionClass.getName());
+                parameters.setParams(configuration.getParams());
+
+                adapterWorkerParametersWorkerSpec.classpath(configuration.getClasspath());
+                adapterWorkerParametersWorkerSpec.setIsolationMode(configuration.getIsolationMode());
+                adapterWorkerParametersWorkerSpec.setDisplayName(configuration.getDisplayName());
+
+                configuration.getForkOptions().copyTo(adapterWorkerParametersWorkerSpec.getForkOptions());
+            }
+        };
+        execute(AdapterWorkerExecution.class, action);
+    }
+
+    @Override
+    public <T extends WorkerParameters> void execute(Class<? extends WorkerExecution<T>> workerExecutionClass, Action<? super WorkerSpec<T>> configAction) {
+        ParameterizedType superType = (ParameterizedType) TypeToken.of(workerExecutionClass).getSupertype(WorkerExecution.class).getType();
+        Class<T> parameterType = Cast.uncheckedNonnullCast(TypeToken.of(superType.getActualTypeArguments()[0]).getRawType());
+        if (parameterType == WorkerParameters.class) {
+            throw new IllegalArgumentException(String.format("Could not create worker parameters: must use a sub-type of %s as parameter type. Use %s for executions without parameters.", ModelType.of(WorkerParameters.class).getDisplayName(), ModelType.of(WorkerParameters.None.class).getDisplayName()));
```

Comment:
Should this be identity equality?  When would this fail?

Reply:
See comment above on equality.  This would fail if you constructed a `WorkerExecution` with `WorkerParameters` as the parameter type.  I've added a test for this.